app/models/throttle.rb

```diff
@@ -54,23 +54,35 @@ def throttled?
 
   def remaining_count
     return 0 if throttled?
-    max_attempts, _attempt_window_in_minutes = Throttle.config_values(throttle_type)
-    max_attempts - attempts
+    Throttle.max_attempts(throttle_type) - attempts
+  end
+
+  def expires_at
+    return Time.zone.now if attempted_at.blank?
+    attempted_at + Throttle.attempt_window_in_minutes(throttle_type).minutes
   end
 
   def expired?
-    return true if attempted_at.blank?
-    _max_attempts, attempt_window_in_minutes = Throttle.config_values(throttle_type)
-    attempted_at + attempt_window_in_minutes.to_i.minutes < Time.zone.now
+    expires_at <= Time.zone.now
   end
 
   def maxed?
-    max_attempts, _attempt_window_in_minutes = Throttle.config_values(throttle_type)
-    attempts >= max_attempts
+    attempts >= Throttle.max_attempts(throttle_type)
+  end
+
+  def self.attempt_window_in_minutes(throttle_type)
+    THROTTLE_CONFIG.with_indifferent_access.dig(throttle_type, :attempt_window)
+  end
+
+  def self.max_attempts(throttle_type)
+    THROTTLE_CONFIG.with_indifferent_access.dig(throttle_type, :max_attempts)
   end
 
-  def self.config_values(throttle_type)
-    config = THROTTLE_CONFIG.with_indifferent_access[throttle_type]
-    [config[:max_attempts], config[:attempt_window]]
+  def self.expires_at(conditions)
+    Enumerator.new do |e|
+      e << find_by(conditions)&.expires_at
+      e << attempt_window_in_minutes(conditions[:throttle_type]) if conditions[:throttle_type]
+      e << 0
+    end.find(&:present?)
```

Comment:
I might be pushing into bad-clever territory here. The general thinking was there's a few different places we can try to get the expiration from, and we'd want to evaluate them lazily.

I suppose a plainer approach is mostly similar now that I think of it...

```suggestion
    record_expires_at = find_by(conditions)&.expires_at
    return record_expires_at if record_expires_at.present?
    return attempt_window_in_minutes(conditions[:throttle_type]) if conditions[:throttle_type]
    0
```

Reply:
Ah, I think the fallback ought to be `Time.zone.now` ? Originally this was implemented as a `throttled_time_remaining` method, returning the minutes 'til the throttle expires.

Reply:
I'm a fan of the expanded if statements, but I also have one more idea for lazy evaluation:

```suggestion
    find_by(conditions)&.expires_at.presence ||
      (conditions[:throttle_type] && attempt_window_in_minutes(conditions[:throttle_type])).presence ||
      0
```

Reply:
In the rebased bd57718, I dropped the class method and just call it from the throttle instance via `Throttle.for`.